CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+
+```
+builtinLogicalComplement [] => []
+
+builtinLogicalComplement [0x0F] => [0xF0]
+
+builtinLogicalComplement [0x4F, 0xF4] => [0xB0, 0x0B]
+```
+
+#### `builtinReadBit`
+
+`builtinReadBit` takes two arguments; we name and describe them below.
+
+1. The `BuiltinByteString` in which the bit we want to read can be found. This
+   is the _data argument_.
+2. A bit index into the data argument, of type `BuiltinInteger`. This is the
+   _index argument_.
+
+Let $b$ refer to the data argument, of length $n$ in bytes, and let $i$ refer to
+the index argument. We use $b[i]$ to refer to the value at index $i$t of $b$; see 
+the [section on the bit indexing scheme](#bit-indexing-scheme) for the exact 
+specification of this.
+
+If $i < 0$ or $i \geq 8 \cdot n$, then `builtinReadBit`
+fails. In this case, the resulting error message must specify _at least_ the
+following information:
+
+* That `builtinReadBit` failed due to an out-of-bounds index argument; and
+* What `BuiltinInteger` was passed as an index argument.
+
+Otherwise, if $b[i] = 0$, `builtinReadBit` returns `False`, and if $b[i] = 1$,
+`builtinReadBit` returns `True`.
+
+Some examples of the intended behaviour of `builtinReadBit` follow. For
+brevity, we write `BuiltinByteString` literals as lists of hexadecimal values.
+
+```
+-- Indexing an empty BuiltinByteString fails
+builtinReadBit [] 0 => error
+
+builtinReadBit [] 345 => error
+
+-- Negative indexes fail
+builtinReadBit [] (-1) => error
+
+builtinReadBit [0xFF] (-1) => error
+
+-- Indexing reads 'from the end'
+builtinReadBit [0xF4] 0 => False
+
+builtinReadBit [0xF4] 1 => False 
+
+builtinReadBit [0xF4] 2 => True 
+
+builtinReadBit [0xF4] 3 => False
+
+builtinReadBit [0xF4] 4 => True
+
+builtinReadBit [0xF4] 5 => True
+
+builtinReadBit [0xF4] 6 => True
+
+builtinReadBit [0xF4] 7 => True
+
+-- Out-of-bounds indexes fail
+builtinReadBit [0xF4] 8 => error
+
+builtinReadBit [0xFF, 0xF4] 16 => error
+
+-- Larger indexes read backwards into the bytes from the end
+builtinReadBit [0xF4, 0xFF] 10 => False 
+```
+
+#### `builtinSetBits`
+
+`builtinSetBits` takes two arguments: we name and describe them below.
+
+1. The `BuiltinByteString` in which we want to change some bits. This is the
+   _data argument_.
+2. A list of index-value pairs, indicating which positions in the data argument
+   should be changed to which value. This is the _change list argument_. Each
+   index has type `BuiltinInteger`, while each value has type `BuiltinBool`.
+
+Let $C = (i_1, v_1), (i_2, v_2), \ldots , (i_k, v_k)$ refer to the change list
+argument, and $b$ the data argument of length $n$ in bytes. Let $b_r$ be the
+result of `builtinSetBits`, whose length is also $n$. We use $b[i]$ to refer to
+the value at index $i$ of $b$ (and analogously, $b_r$); see the [section on the bit
+indexing scheme](#bit-indexing-scheme) for the exact specification of this.
+
+If the change list argument is empty, the result is $b$. If the change list
```

Comment:
I don't think it would be a problem to define the algorithm by recursion. Given the audience for this kind of document, I think that would probably be a clearer way of communicating it!

Reply:
I was kind of going for that exactly, but clearly that didn't work so well. Can you suggest a better way I could have phrased this?